File: bundle-deployer/src/validation/transforms_are_valid.ts
// Validation: Check that transform JSON files are valid

import type { Bundle } from "../types/bundle.ts";
import { getErrorMessage } from "../utils/error.ts";

export async function run(base: string, bundle: Bundle): Promise<void> {
  for (const table of bundle.tables) {
    const transformNames = new Set<string>();
    
    for (const transform of table.transforms) {
      const fullPath = `${base}/${transform.path}`;
      
      // Read the transform file
      const content = await Deno.readTextFile(fullPath);
      
      // Parse as JSON
      let transformJson;
      try {
        transformJson = JSON.parse(content);
      } catch (e) {
        throw new Error(
          `Transform file is not valid JSON: path=${fullPath} error=${getErrorMessage(e)}`
        );
      }
      
      // Check for non-empty name field
      const name = transformJson.name;
      
      if (name === undefined || name === null) {
        throw new Error(
          `Transform file missing required 'name' field: path=${fullPath}`
        );
      }
      
      if (typeof name !== 'string') {
        throw new Error(
          `Transform file 'name' field is not a string: path=${fullPath}`
        );
      }
      
      if (name.trim().length === 0) {
        throw new Error(
          `Transform file has empty 'name' field: path=${fullPath}`
        );
      }
      
      // Check for duplicate transform names
      if (transformNames.has(name)) {
        throw new Error(
          `Duplicated transform name '${name}' path=${fullPath}`
        );
      }
      transformNames.add(name);
      
      // Check subtype if present
      if (transformJson.subtype !== undefined) {
        const subtype = transformJson.subtype;
        
        if (typeof subtype !== 'string') {
          throw new Error(
            `Transform file 'subtype' field is not a string: path=${fullPath}`
          );
        }
        
        if (subtype !== 'firehose') {
          throw new Error(
            `Transform file has invalid subtype '${subtype}', must be 'firehose': path=${fullPath}`
          );
        }
      }
    }
  }
}

File: src/validate/transforms_are_valid.rs
use crate::bundle::Bundle;
use serde_json::Value;
use std::collections::HashSet;
use tokio::fs;

pub async fn run(base: &str, bundle: &Bundle) -> Result<(), String> {
    for table in &bundle.tables {
        let mut transform_names: HashSet<String> = HashSet::new();

        for transform in &table.transforms {
            let full_path = format!("{base}/{}", transform.path);

            // Read the transform file
            let content = match fs::read_to_string(&full_path).await {
                Ok(v) => v,
                Err(e) => {
                    return Err(format!(
                        "ERROR: {}.{} Failed to read transform file: path={full_path} error={e}",
                        file!(),
                        line!()
                    ));
                }
            };

            // Parse as JSON
            let transform_json: Value = match serde_json::from_str(&content) {
                Ok(v) => v,
                Err(e) => {
                    return Err(format!(
                        "ERROR: {}.{} Transform file is not valid JSON: path={full_path} error={e}",
                        file!(),
                        line!()
                    ));
                }
            };

            // Check for non-empty name field
            let this_name = match transform_json.get("name") {
                Some(name_value) => match name_value.as_str() {
                    Some(name_str) => {
                        if name_str.trim().is_empty() {
                            return Err(format!(
                    "ERROR: {}.{} Transform file has empty 'name' field: path={full_path}",
                    file!(),
                    line!()
                ));
                        }
                        name_str // Return the string if valid
                    }
                    None => {
                        return Err(format!(
                "ERROR: {}.{} Transform file 'name' field is not a string: path={full_path}",
                file!(),
                line!()
            ));
                    }
                },
                None => {
                    return Err(format!(
            "ERROR: {}.{} Transform file missing required 'name' field: path={full_path}",
            file!(),
            line!()
        ));
                }
            };

            if transform_names.contains(this_name) {
                return Err(format!(
                    "ERROR: {}.{} Duplicated transform name '{this_name}' path={full_path}",
                    file!(),
                    line!()
                ));
            }
            transform_names.insert(this_name.to_string());

            // Check subtype if present
            if let Some(subtype_value) = transform_json.get("subtype") {
                match subtype_value.as_str() {
                    Some(subtype_str) => {
                        if subtype_str != "firehose" {
                            return Err(format!(
                                "ERROR: {}.{} Transform file has invalid subtype '{}', must be 'firehose': path={full_path}",
                                file!(),
                                line!(),
                                subtype_str
                            ));
                        }
                    }
                    None => {
                        return Err(format!(
                            "ERROR: {}.{} Transform file 'subtype' field is not a string: path={full_path}",
                            file!(),
                            line!()
                        ));
                    }
                }
            }
        }
    }

    Ok(())
}
